### DISCUSSION TITLE
How to check the CPU for support of the required instruction set?

### DISCUSSION THREAD

**Discussion by 3galki** [zlib-ng/zlib-ng#1162] [Q&A]
Is there a function in your library to check whether the CPU supports the required instructions set or not?
Or if I have to do it manually can you provide an example?

**Comment by mtl1979** [comment]
If you statically link against zlib-ng, you can just reuse the existing variables... That however doesn't work if you use the shared version of zlib-ng.

**Comment by 3galki** [comment]
What variables did you mean? I understand how to check which instruction sets are available. But how can I find out which ones were really used in the current zlib-ng build?

**Reply by mtl1979** [reply]
Basically the used functions are stored in the function table structure. When statically linking, you can read the function table, because it doesn't need to be exported. When linking with the shared version, function table is not visible, and as such it is not possible to know what variant is used for each function.

**Comment by 3galki** [comment]
Sorry for my stupid question.  It looks like you choose relevant optimized functions at runtime.
Did I understand right I can hope that zlib-ng will work on any CPU?

**Reply by mtl1979** [reply]
Most of the variants are selected at runtime, but it's possible to modify the build configuration to disable runtime check and use all available instructions.

zlib-ng should work on most little- or big-endian processors no matter if unaligned access is allowed or not... That doesn't mean however that it can be compiled with every possible compiler version. There might be unknown issues with architectures that haven't been tested yet.

Most notably, support for 16-bit processors has been dropped to simplify the code.
